campaigns/workflows.md

```diff
@@ -0,0 +1,345 @@
+  stored at scattered locations unless the download cost is extremely low.)  This
+  would make it possible for a master "config file" be created for `intake` so that
+  an individual catalog need never actually be loaded (i.e., the master catalog is
+  loaded automatically).  This could also make it possible to search and discover
+  datasets directly from within the Notebook, which might eliminate the need for a
+  JupyterLab extension or external website service for search and discovery.
+- Automatic catalog generation for dataset (choose your dataset here)
+
+### Analysis
+
+Data analysis is the process of turning computational data into knowledge.  This
+is done primarily by computing one or many *reductions* of the original data and
+visualizing these reductions.  Sometimes the visualizations are tabular (i.e.,
+just numbers in a table), and sometimes the visualizations are graphical (i.e.,
+generating plots).  Reductions can be *slices* of the original dataset, or they
+can be wholely new computed artifacts.  Reductions are almost always computed
+because the original data usually has more then 2 or 3 dimensions.  Three-dimensional
+rendering of data is certainly possible on 2D screens, but this is only possible
+if the platform allows dynamic interaction with the visualization.  If the
+third dimension is time, then the data can be visualized with a movie, but the
+practicality of that depends on how long the movie would be to watch the entire
+time history.  Normally, a subset of the data is selected, even for 3D rendering
+or animation.
+
+Xarray provides an excellent interface for slicing, subsetting, and quickly
+visualizing 1D or 2D data.  It also provides an excellent interface for basic
+operations on the data to compute new data artifacts (e.g., climatologies,
+anomalies, etc.).  Some artifacts, however, are complicated to produce, and
+may require an Xarray-aware package to provide the new operation.  We refer
+to such abstractions as *Operators*.  At NCAR, new Operators should be placed
+in the GeoCAT package.
+
+Visualization can also be complex, and canned visualization routines for specific
+kinds of plots should be placed in the GeoCAT-viz package.
+
+With the success of Xarray and visualization packages like Matplotlib and
+Cartopy, much of what is needed for analysis already exists, but there are
+analysis operations that are routinely done that are needlessly complex.  These
+operations should be functionalized and packaged.
+
+#### Possible Projects
+
+- Encoding a complete coverage of the CF conventions is too great of a task
+  for a single project, but adding functionality to
+  [CF-Xarray](https://github.com/xarray-contrib/cf-xarray) would make things
+  much easier for users.
+- Better calendar capabilities would be much appreciated.  Traditional
+  timestamps used in Pandas and Xarray are fine but they don't necessarily
+  work with non-standard calendars.  Any time a *time difference* must be
```

Comment:
> Traditional timestamps used in Pandas and Xarray are fine but they don't necessarily
  work with non-standard calendars. 

xarray supports non-standard calendars via cftime. Should it be removed from this statement or can this be updated to reflect this difference between xarray and pandas time handling functionality?

Reply:
What does it support?  Does it support calculating time differences properly?  My understanding is that it only supports converting the `time` coordinate to an equivalent `datetime`-like object, and after that time differencing is done on a proleptic Gregorian calendar.

Reply:
...maybe I'm wrong.  Just looking at the `cftime` API suggests that there should be a way to compute time differences correctly...*and* maybe it can even convert calendars but going through the Julian calendar?  I don't know.  The documentation is very limited.

Reply:
As far as I can tell, xarray is aware of the calendar when performing time differencing.. Let me know if I misunderstood anything


Here's a quick example:


```python
In [43]: t = xr.cftime_range(start="202001", freq="2MS", periods=7, calendar='noleap')

In [44]: ts = xr.DataArray(np.vstack([t[:-1], t[1:]]).T, dims=['time', 'bnds'])

In [45]: ts
Out[45]:
<xarray.DataArray (time: 6, bnds: 2)>
array([[cftime.DatetimeNoLeap(2020-01-01 00:00:00),
        cftime.DatetimeNoLeap(2020-03-01 00:00:00)],
       [cftime.DatetimeNoLeap(2020-03-01 00:00:00),
        cftime.DatetimeNoLeap(2020-05-01 00:00:00)],
       [cftime.DatetimeNoLeap(2020-05-01 00:00:00),
        cftime.DatetimeNoLeap(2020-07-01 00:00:00)],
       [cftime.DatetimeNoLeap(2020-07-01 00:00:00),
        cftime.DatetimeNoLeap(2020-09-01 00:00:00)],
       [cftime.DatetimeNoLeap(2020-09-01 00:00:00),
        cftime.DatetimeNoLeap(2020-11-01 00:00:00)],
       [cftime.DatetimeNoLeap(2020-11-01 00:00:00),
        cftime.DatetimeNoLeap(2021-01-01 00:00:00)]], dtype=object)
Dimensions without coordinates: time, bnds

In [46]: a = ts.diff('bnds')


In [48]: t = xr.cftime_range(start="202001", freq="2MS", periods=7)

In [49]: ts = xr.DataArray(np.vstack([t[:-1], t[1:]]).T, dims=['time', 'bnds'])

In [50]: ts
Out[50]:
<xarray.DataArray (time: 6, bnds: 2)>
array([[cftime.DatetimeGregorian(2020-01-01 00:00:00),
        cftime.DatetimeGregorian(2020-03-01 00:00:00)],
       [cftime.DatetimeGregorian(2020-03-01 00:00:00),
        cftime.DatetimeGregorian(2020-05-01 00:00:00)],
       [cftime.DatetimeGregorian(2020-05-01 00:00:00),
        cftime.DatetimeGregorian(2020-07-01 00:00:00)],
       [cftime.DatetimeGregorian(2020-07-01 00:00:00),
        cftime.DatetimeGregorian(2020-09-01 00:00:00)],
       [cftime.DatetimeGregorian(2020-09-01 00:00:00),
        cftime.DatetimeGregorian(2020-11-01 00:00:00)],
       [cftime.DatetimeGregorian(2020-11-01 00:00:00),
        cftime.DatetimeGregorian(2021-01-01 00:00:00)]], dtype=object)
Dimensions without coordinates: time, bnds

In [51]: b = ts.diff('bnds')

In [52]: a # DatetimeNoLeap calendar
Out[52]:
<xarray.DataArray (time: 6, bnds: 1)>
array([[5097600000000000],
       [5270400000000000],
       [5270400000000000],
       [5356800000000000],
       [5270400000000000],
       [5270400000000000]], dtype='timedelta64[ns]')
Dimensions without coordinates: time, bnds

In [53]: b # DatetimeGregorian calendar
Out[53]:
<xarray.DataArray (time: 6, bnds: 1)>
array([[5184000000000000],
       [5270400000000000],
       [5270400000000000],
       [5356800000000000],
       [5270400000000000],
       [5270400000000000]], dtype='timedelta64[ns]')
Dimensions without coordinates: time, bnds
```

Reply:
I added some text in the latest commit to address this.  There are actually 2 things that need to be done:

1. correct computation of time durations (differences in datetimes), and
2. conversion from one calendar to another.

If `cftime` answers (1), then we need to test to see if it also answers (2).  If so, then we might need to add that functionality into `pint` so that it can convert units properly for times in different calendars.